docs/architecture/adr-versions.md

```diff
@@ -0,0 +1,219 @@
+# Protocol and Chain Versions
+
+The Software Version is covered by SemVer and described elsewhere.
+It is not relevant to the protocol description, suffice to say that if any protocol version
+changes, the software version changes. Software version is included in NodeInfo for convenience/diagnostics.
+
+Here we focus on protocol and chain versions.
+
+## Requirements
+
+We need to version blockchains across protocols, networks, forks, etc.
+We need to do it in a way that is scalable and maintainable - we can't just litter
+the code with conditionals.
+We need chain identifiers and descriptions so we can talk about a multitude of chains in a meaningful way.
+
+Let's start with versions.
+
+### Protocol Version
+
+The complete version of the protocol contains the following sub-versions:
+BlockVersion, P2PVersion, AppVersion. These versions reflect the major sub-components
+of the software that are likely to evolve together, at different rates, and in different ways,
+as described below.
+
+#### BlockVersion
+
+- All tendermint hashed data-structures (headers, votes, txs, responses, etc.).
+	- Note the semantic meaning of a transaction may change according to the AppVersion,
+		but the way txs are merklized into the header is part of the BlockVersion
+- It should be the least frequent/likely to change.
+	- Tendermint should be stabilizing - it's just Atomic Broadcast.
+	- We can start considering for Tendermint v2.0 in a year
+- For now, assume that BlockVersion will not change in the life of a given chain, except by critical emergency.
+- To effectively change it, dump state and start a new blockchain with the new BlockVersion.
+- It's easy to determine the version of a block from its serialized form
+
+#### P2PVersion
+
+- All peer messaging (messages, detectable behaviour)
+- Will change gradually as reactors evolve to improve performance and support new features
+	- eg proposed new message types BatchTx in the mempool and HasBlockPart in the consensus
+- It's easy to determine the version of a peer from its first serialized message/s
+- New versions must be compatible with at least one old version to allow gradual upgrades
```

Comment:
I don't understand this. Does this mean that the changes between two versions should always be done in a backward compatible way.

Reply:
I'd prefer we not require compatibility with old versions. That seems kind of unnatural to me, plus its unenforceable. I think listing all the previous versions a node supports would be more ideal.

Reply:
> Does this mean that the changes between two versions should always be done in a backward compatible way.

For the p2p layer, yes. It means you can upgrade at least once and still connect successfully to the rest of the network.

Not sure why we wouldn't want that requirement at the peer layer. 

> I think listing all the previous versions a node supports would be more ideal.

This is what we're proposing, and requiring the list be non-empty